### PR TITLE
Dev Containers

### LINKED ISSUE #539: Rework or add an env var for hard coded service names
Context: so I have this habit of naming containers and services with a naming convention that groups multiple services by their project name and then the image type. For example, _cleanslate-graphql_, _cleanslate-postgres_, _cleanslate-authentication_, _cleanslate-client_ etc. 

In the case of Clean Slate I updated the service name and the container_name as such. I do that so I know what container belongs to what project and networks end up like _cleanslate_default_ etc. It really helps when removing containers to know I'm not removing the wrong container; deleting the wrong database, for example.

Now, since I'm running Caddy outside the Clean Slate docker-compose and on its own, this in case I want to use it for other projects, I followed the [Caddy Docker docs](https://caddyserver.com/docs/running#docker-compose) where it mentions the following:

> If you're using Caddy to [reverse proxy](https://caddyserver.com/docs/caddyfile/directives/reverse_proxy) to another container, remember that in Docker networking, localhost means "this container", not "this machine". So for example, do not use reverse_proxy localhost:8080, instead use reverse_proxy other-container:8080

That's fine. I just update the Caddyfile and changed `localhost:port` to `service_name:port`. For example, from `localhost:8080` to `cleanslate-graphql:8080`. I add in caddy_default network to the the Clean Slate services that need it, so they can talk with each other, and everything is fine in the world. I launch everything and all appears _a-ok_.

Of course, your docs don't mention anything about installing Caddy in the same compose "stack" to take advantage of localhost that was generated in the Caddyfile, via the configuration.sh file, and it also doesn't say anything about making sure _not_ to change service names. So I go about my business and everything is working great, except I can't log in with the newly created profile's apiToken. I just get the popup that says to use the apiToken instead of the authId. Umm, I know I'm using apiToken, thank you very much lol.

Well, I have access to console, I can see the row is created in the profile table, I can see it's created in the database directly, I can see the login screen, I can hit the /healthz route, so on and so forth. But I can't log in? 

Anyway, after many detective hours later, trying to figure out why everything is working, but I'm unable to login, I discover this:

```typescript
const graphqlUrl = isProduction
  ? // We need server to server communication (over HTTP) here.
    // In other words, the authentication server container to Hasura container on production.
    // And the authentication server to Hasura container on development.
    // As the authentication server is not in a container on development.
    // If you have the authentication server go through Caddy to the Hasura server, it will strip the X-Hasura-Role
    // This is usually exactly what you want for security! However, for the /auth/graphql route
    // It is a problem, as stripping out X-Hasura-Role will have Hasura default to the admin role
    // Simply because you are also passing X-Hasura-Admin-Secret at the same time.
    // In this scenario, you would have the request made with admin permissions, effectively.
    'http://graphql-server:8080/v1/graphql'
  : 'http://localhost:8080/v1/graphql'
```
https://github.com/successible/cleanslate/blob/main/src/server.ts#L30-L41

The service name has been hard coded into the server.ts file. So I change the graphql service name back to `graphql-server` (and I do the authentication-server just in case as well, but I didn't find it hard coded in my search) and instantly, I'm able to log in. So yeah, the service name needing to be set exactly, since it's hardcoded, was the issue.

The request: all that being said, I'm kindly asking to have that either reworked somehow or replaced with an env var so an override env var can be added to the .env file. If nothing else, maybe the docs could be updated with a note about changing service names.

Aside: maybe something should be updated in the configuration.sh file to ask if Caddy is within the stack or outside standalone. That way it can update the Caddyfile accordingly, by changing locahost to the service names etc.

_Sorry for the long drawn out issue, but yeah..._

### LINKED ISSUE #551: Dev Containers
I am hoping to help out and contribute to the repo! 

In my development workflow, I prefer to keep things isolated and "containerized" so I don't have to worry about having any/all types of dev dependencies installed.

Are you open to supporting dev containers or specific dev dockerfile to use in tandem with `docker-compose-dev.yaml`?

### PR DESCRIPTION
**Pull request by McMumf**
## What did you do and why?

Created and configured a dev container to help standardize the development environment for contributing back to this repo.

I also started adding unit tests. I can probably start adding more if you want me to make another issue where we can discuss.

To better support this, I also had to start implementations for #539 

closes #539
closes #551

### CODE REVIEW

**Comment by paulzakin** [comment]
> I also started adding unit tests. I can probably start adding more if you want me to make another issue where we can discuss.

Might as well add them on this same PR :)

Gave it an initial look - seems OK to me - but have to read up a bit more on Dev containers - never used them personally.

**Comment by McMumf** [comment]
Selfishly, I have linux on my main computer and Mac for my laptop so making sure all the tools are installed and configured properly with other things I have running was getting to be a pain.

So dev containers help me keep things segregated.

I tried to stick to the dev workflow you have outlined in the README and `dev.sh` when building the dev container. Otherwise, it's left mostly alone with docker compose overrides.

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+RUN apt-get update && apt-get install -y \
+    jq \
+    curl \
+    unzip \
+    ca-certificates \
+    gnupg2 \
+    && rm -rf /var/lib/apt/lists/*
+
+# Docker CLI
+RUN curl -fsSL https://download.docker.com/linux/static/stable/x86_64/docker-27.5.1.tgz \
```
Comment (paulzakin) [inline .devcontainer/Dockerfile]: Anyway to have a latest - rather than pinning here - it is easier to maintain for me

File: .devcontainer/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,55 @@
+FROM mcr.microsoft.com/devcontainers/typescript-node:24-trixie
```
Comment (paulzakin) [inline .devcontainer/Dockerfile:1]: Anyway to have a latest - rather than pinning here - it is easier to maintain for me

File: src/server.ts [hunk 3]
```diff
@@ -27,18 +27,18 @@ app.use(logger(isDevelopment ? { transport: { target: 'pino-pretty' } } : {}))
...
-  ? // We need server to server communication (over HTTP) here.
-    // In other words, the authentication server container to Hasura container on production.
-    // And the authentication server to Hasura container on development.
-    // As the authentication server is not in a container on development.
-    // If you have the authentication server go through Caddy to the Hasura server, it will strip the X-Hasura-Role
-    // This is usually exactly what you want for security! However, for the /auth/graphql route
-    // It is a problem, as stripping out X-Hasura-Role will have Hasura default to the admin role
-    // Simply because you are also passing X-Hasura-Admin-Secret at the same time.
-    // In this scenario, you would have the request made with admin permissions, effectively.
-    'http://graphql-server:8080/v1/graphql'
```
Comment (paulzakin) [inline src/server.ts:40]: I would prefer it still have this code and just default to the hasura endpoint when available, like you did with NEXT_PUBLIC_AUTH_HOST. After all, production systems already running will not have this value defined.

**Review by paulzakin — CHANGES_REQUESTED** [review]

File: .devcontainer/Dockerfile (hunk 2 above)
Comment (McMumf) [reply .devcontainer/Dockerfile:1]: I'll have to research the behavior of dev containers when using a latest tag. I'm sure there's some way to auto magically rebuild it using the existing volumes if the digest/sha has changed.

Comment (paulzakin) [reply .devcontainer/Dockerfile:1]: That would be great!

Comment (McMumf) [reply .devcontainer/Dockerfile:1]: I don't see anything about pinning to latest and automatically rebuilding.

However, renovate may be of use here! [Renovate supports updating [Devcontainer](https://code.visualstudio.com/docs/devcontainers/containers) dependencies.](https://docs.renovatebot.com/modules/manager/devcontainer/).

This might be the best option, otherwise, it seems the intended workflow is to rebuild and reopen.

Also, the devcontainer image doesn't have a latest tag: [Typescript-Node Devcontainer Repos and Tags](https://mcr.microsoft.com/en-us/artifact/mar/devcontainers/typescript-node/tags). We would have to have it be a general latest.

I suggest we make use of renovate. I understand wanting to use latest, however, there are some issues when a new LTS has breaking changes. Pinning to a specific LTS gives time to test out and fix anything on a newer LTS.

File: src/server.ts (hunk 3 above)
Comment (McMumf) [reply src/server.ts:40]: I believe I have addressed this. I will leave it up to you to resolve the comment just in case you had something else in mind.

File: .devcontainer/Dockerfile (hunk 1 above)
Comment (McMumf) [reply .devcontainer/Dockerfile]: Yes, since the base container is debian based, I used the recommended installation method for that.

I also updated the file mount so that the user's (in this case `node`) home directory across rebuilds so things like IDE specific configs and pnpm cache is saved across docker rebuilds. Helpful if contributors want to use git from within the container as well.

File: .devcontainer/Dockerfile (hunk 2 above)
Comment (paulzakin) [reply .devcontainer/Dockerfile:1]: That’s fine, we use renovate currently, we’ll see if it picks it up

File: .devcontainer/Dockerfile (hunk 1 above)
Comment (paulzakin) [reply .devcontainer/Dockerfile]: Ya this seems good

File: src/server.ts (hunk 3 above)
Comment (paulzakin) [reply src/server.ts:40]: Maybe I’m missing something, I see getGraphqlUrl but it doesn’t seem like you have invoked it in server.ts - still seems that it is calling the environment variable directly

**Review by paulzakin — CHANGES_REQUESTED** [review]

File: .devcontainer/devcontainer.json [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    "3001": {
+      "label": "Auth server"
+    }
+  },
+  "customizations": {
+    "vscode": {
+      "extensions": [
+        "dbaeumer.vscode-eslint",
+        "biomejs.biome",
+        "bradlc.vscode-tailwindcss",
```
Comment (paulzakin) [inline .devcontainer/devcontainer.json]: bradlc.vscode-tailwindcss  Probably should be removed as we don’t use tailwind

File: .devcontainer/README.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
+# Dev Container
+
+## Setup
```
Comment (paulzakin) [inline .devcontainer/README.md]: Can you move this to the main read me in the local development section?

**Review by paulzakin — CHANGES_REQUESTED** [review]

File: src/server.ts (hunk 3 above)
Comment (McMumf) [reply src/server.ts:40]: Someone may have forgotten to `git add` a file...

I'll get to it later this week

File: .devcontainer/README.md (hunk 5 above)
Comment (McMumf) [reply .devcontainer/README.md]: Please let me know if the location should be changed

**Comment by paulzakin** [comment]
This seems good to me - anything else you want to add?

**Comment by paulzakin** [comment]
action failed because server.ts is weird - it cannot define import from the rest of clean slate - so define getGraphqlUrl in that file and import it in the rest of the system

**Comment by McMumf** [comment]
I don't think I can see the actions. `import { getGraphqlUrl } from './helpers/getGraphqlUrl'` is defined on line 6 of `server.ts`. 

I ran `cd src && npx tsc && npx biome check && npx next info && npx next build --webpack` to see if that would have a similar error

```shell
$ cd src && npx tsc && npx biome check && npx next info && npx next build --webpack
Checked 347 files in 114ms. No fixes applied.

Operating System:
  Platform: linux
  Arch: arm64
  Version: #1 SMP PREEMPT_DYNAMIC Thu Oct 10 22:53:48 UTC 2024
  Available memory (MB): 7359
  Available CPU cores: 6
Binaries:
  Node: 24.18.0
  npm: 11.16.0
  Yarn: 1.22.22
  pnpm: 11.9.0
Relevant Packages:
  next: 16.2.10 // There is a newer version (16.2.11) available, upgrade recommended!
  eslint-config-next: N/A
  react: 19.2.7
  react-dom: 19.2.7
  typescript: 6.0.3
Next.js Config:
  output: export
⚠ There is a newer version (16.2.11) available, upgrade recommended!
   Please try the latest canary version (`npm install next@canary`) to confirm the issue still exists before creating a new issue.
   Read more - https://nextjs.org/docs/messages/opening-an-issue
⚠ You are using a non-standard "NODE_ENV" value in your environment. This creates inconsistencies in the project and is strongly advised against. Read more: https://nextjs.org/docs/messages/non-standard-node-env
Attention: Next.js now collects completely anonymous telemetry regarding usage.
This information is used to shape Next.js' roadmap and prioritize features.
You can learn more, including how to opt-out if you'd not like to participate in this anonymous program, by visiting the following URL:
https://nextjs.org/telemetry

▲ Next.js 16.2.10 (webpack)

✓ Finished TypeScript in 3.3s
  Creating an optimized production build ...
✓ Compiled successfully in 6.8s
✓ Collecting page data using 5 workers in 2.1s
Error: <Html> should not be imported outside of pages/_document.
Read more: https://nextjs.org/docs/messages/no-document-import-in-page
    at z (.next/server/chunks/646.js:6:1323)
Error occurred prerendering page "/404". Read more: https://nextjs.org/docs/messages/prerender-error
Error: <Html> should not be imported outside of pages/_document.
Read more: https://nextjs.org/docs/messages/no-document-import-in-page
    at z (.next/server/chunks/646.js:6:1323)
Export encountered an error on /404, exiting the build.
⨯ Next.js build worker exited with code: 1 and signal: null
```

This is from inside the dev container

**Comment by paulzakin** [comment]
All you have to do is just take that helper function and stick it in that server.ts file, and then import that elsewhere if that makes sense. Basically, due to the way the type script configuration worked here, the server.ts cannot good import code from the rest of clean slate

**Comment by paulzakin** [comment]
Almost there - just gotta run the lint command in package.json to fix a formatting error

**Comment by paulzakin** [comment]
About ready to merge - do the dev containers work okay locally?

**Comment by McMumf** [comment]
Yep! I was able to verify login and some basic interactions work as expected.

My next GitHub issue and suggestion would be automated tests so I know I didn't inadvertently break any nuanced edge cases or some basic functionality I didn't know existed/don't use.

**Review by paulzakin — APPROVED** [review]

**Comment by paulzakin** [comment]
Nice work, merged!
